csharp/extractor/Semmle.Extraction.CSharp.Standalone/BuildAnalysis.cs

```diff
@@ -42,20 +46,18 @@ interface IBuildAnalysis
     /// <summary>
     /// Main implementation of the build analysis.
     /// </summary>
-    class BuildAnalysis : IBuildAnalysis
+    class BuildAnalysis : IBuildAnalysis, IDisposable
     {
         private readonly AssemblyCache assemblyCache;
         private readonly NugetPackages nuget;
         private readonly IProgressMonitor progressMonitor;
-        private HashSet<string> usedReferences = new HashSet<string>();
-        private readonly HashSet<string> usedSources = new HashSet<string>();
-        private readonly HashSet<string> missingSources = new HashSet<string>();
-        private readonly Dictionary<string, string> unresolvedReferences = new Dictionary<string, string>();
+        private readonly IDictionary<string, bool> usedReferences = new ConcurrentDictionary<string, bool>();
```

Comment:
Perhaps we should add this to Semmle.Util and use instead:
```
using System.Collections;
using System.Collections.Concurrent;
using System.Collections.Generic;

namespace Semmle.Util
{
    public class ConcurrentHashSet<T> : IEnumerable<T>
    {
        private readonly ConcurrentDictionary<T, object> dict;
        public ConcurrentHashSet()
        {
            dict = new ConcurrentDictionary<T, object>();
        }

        public void Add(T t) => dict[t] = null;

        public bool Contains(T t) => dict.ContainsKey(t);

        public IEnumerator<T> GetEnumerator() => dict.Keys.GetEnumerator();

        IEnumerator IEnumerable.GetEnumerator() => dict.Keys.GetEnumerator();
    }
}
```

Reply:
My reservation with this is that this would be a half-baked implementation, that should also implement
```
[System.Serializable]
public class ConcurrentHashSet<T> : System.Collections.Generic.ICollection<T>, System.Collections.Generic.IEnumerable<T>, System.Collections.Generic.IReadOnlyCollection<T>, System.Collections.Generic.ISet<T>, System.Runtime.Serialization.IDeserializationCallback, System.Runtime.Serialization.ISerializable
```
So we should really just reference the nuget package `ConcurrentHashSet` which I presume is totally correct and bug-free.....

Then it means updating resources/lib which is a bit of a pain.

Reply:
I think a half-baked implementation suffices, as we only need to add the things that we actually need. If we ever need, say, to use it as an `ICollection<T>`, then we can implement it then.

Reply:
There's literally one place where we use `ConcurrentDictionary` instead of `ConcurrentHashSet` so I'm really not convinced it's worth the effort. Particularly because we'd need a unit test suite for the new `ConcurrentHashSet` class.

Reply:
OK 👍